drivers/clock_control/clock_control_nrf2_hfxo.c

```diff
@@ -68,13 +102,52 @@ static void onoff_start_hfxo(struct onoff_manager *mgr, onoff_notify_fn notify)
 	k_timer_start(&dev_data->timer, dev_config->start_up_time, K_NO_WAIT);
 }
 
-static void onoff_stop_hfxo(struct onoff_manager *mgr, onoff_notify_fn notify)
+static void stop_hfxo(const struct device *dev)
 {
-	struct dev_data_hfxo *dev_data =
-		CONTAINER_OF(mgr, struct dev_data_hfxo, mgr);
+	struct dev_data_hfxo *dev_data = dev->data;
 
 	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_STOPREQHFXO);
 	soc_lrcconf_poweron_release(&dev_data->hfxo_node, NRF_LRCCONF_POWER_MAIN);
+}
+
+static void release_hfxo(const struct device *dev)
+{
+	struct dev_data_hfxo *dev_data = dev->data;
+	unsigned int key;
+
+	key = irq_lock();
+	if (dev_data->requested && dev_data->req_count_zli == 0) {
+		stop_hfxo(dev);
+	}
+	dev_data->requested = false;
+	irq_unlock(key);
+}
+
+static void release_zli_hfxo(const struct device *dev)
+{
+	struct dev_data_hfxo *dev_data = dev->data;
+	unsigned int key;
+
+	key = irq_lock();
+	if (dev_data->req_count_zli < 1) {
+		irq_unlock(key);
+		/* Early return due to no requests pending */
+		return;
+	}
+
+	dev_data->req_count_zli--;
+	if (dev_data->req_count_zli < 1 && !dev_data->requested) {
+		stop_hfxo(dev);
+	}
+
+	irq_unlock(key);
+}
+
+static void onoff_stop_hfxo(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	const struct device *dev = DEVICE_DT_INST_GET(0);
```

Comment:
You don't need to do this. `start_hfxo`, `stop_hfxo`, `request_*` and `release_*` can take `struct dev_data_hfxo *dev_data` as the parameter and this value can be obtained here with `CONTAINER_OF()`.

Reply:
Changed to make it similar to `onoff_start_hfxo`.